net: conn_mgr: documentation and implementation guidelines

glarsennordic opened the pull request.

glarsennordic committed:
net: conn_mgr: Write documentation

Write documentation for conn_mgr and its subsystems, especially
guidelines for writing connectivity implementations

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

glarsennordic commented:
Not yet proofread, not yet ready for review

glarsennordic commented:
There are still a couple minor crosslink errors, and I need to update codeowners / maintainers files to include the new docs, but otherwise these docs should be ready for review now

jukkar commented on doc/connectivity/networking/conn_mgr/main.rst:
```diff
@@ -0,0 +1,441 @@
+           /* Otherwise, it's some other event type we didn't register for. */
+   }
+
+   int main() {
```
Please place { to next line

glarsennordic commented on doc/connectivity/networking/conn_mgr/implementation.rst:
```diff
@@ -0,0 +1,471 @@
+.. code-block:: c
+
+   /* Create the API implementation functions */
+   int my_connect_impl(struct conn_mgr_conn_binding *const binding) {
+	/* Your implementation goes here */
+   }
+   int my_disconnect_impl(struct conn_mgr_conn_binding *const binding) {
+	/* Your implementation goes here */
+   }
+   void my_init_impl(struct conn_mgr_conn_binding *const binding) {
+	/* Your implementation goes here */
+   }
```
@jukkar Picking up what you mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/62126/commits/1afe35927b6d5fa7cc3ee9c3f0cb7725e3389a6b#r1317146403

Would writing a bit more detail in these comments satisfy what you are requesting?

jukkar replied:
Perhaps. It would be nice to see some info what user could do with the binding, it is not clear now.

glarsennordic replied:
Okay, I've updated the comments to be a bit more specific. But I want to avoid delving into the specifics of what an implementation is expected to do since that is covered by the other sections of this document, and it would distract from the main point of this section

jukkar commented on doc/connectivity/networking/conn_mgr/main.rst:
```diff
@@ -0,0 +1,442 @@
+   /* Declare a net_mgmt callback struct to store the callback */
+   struct net_mgmt_event_callback my_conn_evt_callback;
+
+   /* Declare a handler to recieve control events */
```
s/recieve/receive/

jukkar commented on doc/connectivity/networking/conn_mgr/main.rst:
```diff
@@ -0,0 +1,442 @@
+            */
+
+           net_mgmt_init_event_callback(
+                   &conn_mgr_conn_callback, conn_mgr_conn_handler,
+				           NET_EVENT_CONN_IF_TIMEOUT | NET_EVENT_CONN_IF_FATAL_ERROR
```
Indentation looks funny here, I suggest indenting with spaces in this example

glarsennordic replied:
Argh, thanks, did my best to get rid of tabs but looks like I missed a line

jukkar commented on include/zephyr/net/conn_mgr_connectivity.h:
```diff
@@ -231,23 +233,27 @@ int conn_mgr_if_get_timeout(struct net_if *iface);
+ */
+
+/**
+ * @brief Connection Manager Bulk versions of select functions from the Connection Manager Connectivity API
```
Too long line, please split to two

wbober commented on doc/connectivity/networking/conn_mgr/main.rst:
```diff
@@ -0,0 +1,442 @@
+
+.. _conn_mgr_monitoring_ignoring_ifaces:
+
+Ignoring ifaces
```
Perhaps add a comment when this is useful?

jukkar commented:
Please check the compliance issues which are currently shown at
https://github.com/zephyrproject-rtos/zephyr/actions/runs/6103554259/job/16564154813?pr=62126

glarsennordic commented on doc/connectivity/networking/conn_mgr/main.rst:
```diff
@@ -0,0 +1,442 @@
+
+   See :ref:`conn_mgr_impl_guidelines_iface_state_reporting` for details.
+
+A ready iface ceases to be ready if any of these conditions is lost.
```
Nope, I meant any of the above conditions -- I'll edit to specify above

glarsennordic commented on doc/connectivity/networking/conn_mgr/main.rst:
```diff
@@ -0,0 +1,442 @@
+Otherwise, the iface will not attempt to reconnect.
+
+.. note::
+   Persistence does affect how connection attempts behave.
```
Ah yes, thank you, I was indeed missing `not`

glarsennordic commented on doc/connectivity/networking/conn_mgr/implementation.rst:
```diff
@@ -0,0 +1,471 @@
+
+Rather than implement all features centrally, Connection Manager relies on each connectivity implementation to implement many behaviors and features individually.
+
+This approach allows Connection Manager to remain lean, and it allows each connectivity implementation to choose the most appropriate approach to these behaviors for itself.
+However it relies on trust that all connectivitly implementations will faithfully implement the features that have been delegated to them.
```
Conectivitly😂

Good catch, thanks

glarsennordic committed:
MAINTAINERS: Update maintainers/codeowners for conn_mgr docs

conn_mgr now has docs, so add them to the maintainers/codeowners files.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

gmarull commented on doc/zephyr.doxyfile.in:
```diff
@@ -2370,6 +2370,7 @@ PREDEFINED             = __DOXYGEN__ \
                          CONFIG_NET_L2_ETHERNET_MGMT \
                          CONFIG_NET_MGMT_EVENT \
                          CONFIG_NET_SOCKETS_POSIX_NAMES \
+                         CONFIG_NET_CONNECTION_MANAGER \
```
please use `__DOXYGEN__` instead (see some other headers)

glarsennordic replied:
If I use `__DOXYGEN__`, then hyperlinks to some of the KConfigs do not work (because those KConfigs are not enabled during docs builds)

I'll try one more time to make sure this was not a fluke, but if not, how do you recommend I fix this issue if not by appending to the doxyfile?

glarsennordic replied:
I was able to get it to work by including a __DOXYGEN__clause in the CMakeLists.txt file

For whatever reason, my local docs builds show errant behavior, so I will have to force-push a build without that addition just to confirm it is truly necessary.

glarsennordic replied:
IDK. I removed it from the CMakeLists.txt, and the KConfig references still work. I'm not sure anymore what problem I was actually experiencing.

rlubos commented on doc/connectivity/networking/conn_mgr/main.rst:
```diff
@@ -0,0 +1,446 @@
+
+Afterwards, ifaces can become ready or unready without firing additional events, so long as there always remains at least one ready iface.
+
+When there are no longer any ready ifaces left, the :c:macro:`NET_EVENT_L4_CONNECTED` :ref:`network management <net_mgmt_interface>` event is triggered, and IP connectivity is said to be unready.
```
It should be `NET_EVENT_L4_DISCONNECTED` here I believe.

rlubos commented on doc/connectivity/networking/conn_mgr/main.rst:
```diff
@@ -0,0 +1,446 @@
+For such ifaces, connectivity control can provide a generic API to request network association (:c:func:`conn_mgr_if_connect`) and dissasociation (:c:func:`conn_mgr_if_disconnect`).
+Network interfaces implement support for this API by :ref:`binding themselves to a connectivity implementation <conn_mgr_impl_binding>`.
+
+Using this API, applications can associate with networks with minimal technology-specific boiler-plate.
```
`boiler-plate` or `boilerplate`?

rlubos commented on doc/connectivity/networking/conn_mgr/main.rst:
```diff
@@ -0,0 +1,446 @@
+Timeout
+-------
+
+The :c:macro:`NET_EVENT_CONN_CMD_IF_TIMEOUT` event is raised when an :ref:`iface association <conn_mgr_control_operation_connecting>` attempt :ref:`times out <conn_mgr_control_timeouts>`.
```
`NET_EVENT_CONN_IF_TIMEOUT`?

rlubos commented on include/zephyr/net/conn_mgr_connectivity.h:
```diff
@@ -39,18 +39,26 @@ extern "C" {
 
+/** @endcond */
+
 enum net_event_ethernet_cmd {
```
The type name should be updated, looks like a copy-paste error.

glarsennordic replied:
Good catch! I'm baffled that this never created problems in any of our tests

rlubos reviewed (commented):
Thanks, noticed a few nits when reading the docs, but overall looks good :+1:

glarsennordic committed:
net: conn_mgr: Fix connectivity event enum type

Connectivity event enum type was erroneously named
net_event_ethernet_cmd.

This PR corrects the name to net_event_conn_cmd.

Signed-off-by: Georges Oates_Larsen <georges.larsen@nordicsemi.no>

rlubos reviewed (approved):
LGTM

carlescufi commented:
All comments by @peknis addressed

kartben commented:
I think someone will want to update the API overview page to point to `conn_mgr_docs` target? Thanks! (and sorry that I always seem to be 15 minutes behind, as of late)

<img width="643" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/128251/153a364f-f9b3-4918-8a45-de384d3fbc7d">

rlubos commented:
@kartben I've opened a PR with the link update: https://github.com/zephyrproject-rtos/zephyr/pull/63233

glarsennordic commented:
Sorry about that @kartben I completely spaced out on updating that. Thanks for pointing it out, and thanks @rlubos for putting a PR in to do it

kartben commented:
> Sorry about that @kartben I completely spaced out on updating that. Thanks for pointing it out, and thanks @rlubos for putting a PR in to do it

No worries at all!
